core/iwasm/libraries/lib-wasi-threads/stress_test/mutex_common.h

```diff
@@ -0,0 +1,261 @@
+    assert(pthread_mutex_unlock(mutex) == 0
+           && "Main thread should be able to unlock a mutex");
+}
+
+void *
+inc_until_limit(void *arg)
+{
+    MutexCounter mutex_counter = *(MutexCounter *)(arg);
+    int sleep_us = 0;
+    while (!pthread_mutex_lock(mutex_counter.mutex)
+           && *mutex_counter.counter < mutex_counter.limit) {
+        (*mutex_counter.counter)++;
+        add_to_stat((int)(pthread_self()));
+        if (mutex_counter.is_sleeping) {
+            sleep_us = hash(*mutex_counter.counter) % 1000;
+        }
+
+        assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+               && "Should be able to unlock a mutex");
+        if (mutex_counter.is_sleeping) {
+            usleep(sleep_us);
+        }
+    }
+
+    assert(*mutex_counter.counter == mutex_counter.limit);
+    assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+           && "Should be able to unlock the mutex after test execution");
+
+    return NULL;
+}
+
+void
+non_main_test(pthread_mutex_t *mutex, int num_iter)
+{
+    MutexCounter mutex_counter;
+    int counter = 0;
+    mutex_counter.mutex = mutex;
+    mutex_counter.counter = &counter;
+    mutex_counter.limit = num_iter;
+    mutex_counter.is_sleeping = false;
+
+    pthread_t tid = 0;
+    spawn_thread(&tid, inc_until_limit, &mutex_counter);
+
+    assert(tid != 0 && "TID can't be 0 after successful thread creation");
+    pthread_join(tid, NULL);
+    assert(*mutex_counter.counter == mutex_counter.limit);
+}
+
+void
+main_non_main_test(pthread_mutex_t *mutex, int num_iter)
+{
+    MutexCounter mutex_counter;
+    int counter = 0;
+    mutex_counter.mutex = mutex;
+    mutex_counter.counter = &counter;
+    mutex_counter.limit = num_iter;
+    mutex_counter.is_sleeping = false;
+
+    pthread_t tid = 0;
+    spawn_thread(&tid, inc_until_limit, &mutex_counter);
+
+    assert(tid != 0 && "TID can't be 0 after successful thread creation");
+    inc_until_limit(&mutex_counter);
+    assert(pthread_join(tid, NULL) == 0
+           && "Thread should be joined without errors");
+    ;
+    assert(*mutex_counter.counter == mutex_counter.limit);
+}
+
+void
+concurrent_inc_test(pthread_mutex_t *mutex, int num_iter, int threads_num,
+                    bool is_sleeping)
+{
+    MutexCounter mutex_counter;
+    int counter = 0;
+    mutex_counter.mutex = mutex;
+    mutex_counter.counter = &counter;
+    mutex_counter.limit = num_iter;
+    mutex_counter.is_sleeping = is_sleeping;
+
+    pthread_t tids[threads_num];
+    for (int i = 0; i < threads_num; ++i) {
+        spawn_thread(&tids[i], inc_until_limit, &mutex_counter);
+    }
+
+    inc_until_limit(&mutex_counter);
+
+    for (int i = 0; i < threads_num; ++i) {
+        assert(pthread_join(tids[i], NULL) == 0
+               && "Thread should be joined without errors");
+        ;
+    }
+}
+
+// This function just runs all the tests described above
+void
+run_common_tests(pthread_mutex_t *mutex)
```

Comment:
we can keep all the tests in one function; splitting them to separate small wasm applications would improve the reporting though.

Reply:
it's a bit controversial to me 

as I'm going to have other mutex types running the same test it might eventually become 

NUM_OF_MUTEX_TYPES x NUM_OF_TESTS matrix what will be hard to update etc

Reply:
well, you will if you execute it on iwasm or other way (not via test-runner) because of stderr logs 

you only won't know if you execute via test-runner and I find it annoying that it does not shown stderr output by the way

Reply:
ok, let's keep it like that. Btw stderr and stdout is shown when the test fails in the test runner IIRC